addons/point_of_sale/static/src/app/services/pos_store.js

```diff
@@ -2715,6 +2720,275 @@ export class PosStore extends WithLazyGetterTrap {
+            const pid = l.product_id.id;
+            if (!productInOrder[pid]) {
+                productInOrder[pid] = {};
+            }
+            productInOrder[pid][l.uuid] = l.qty;
+        });
+        const matchingCombos = [];
+        for (const comboProduct of this.models["product.template"].filter(
+            (p) => p.type == "combo"
+        )) {
+            const combos = comboProduct.combo_ids;
+            const availableQty = JSON.parse(JSON.stringify(productInOrder));
+            const quantityTaken = {};
+            let comboQty = Infinity;
+
+            for (const combo of combos) {
+                if (combo.is_upsell) {
+                    continue;
+                }
+                const comboId = combo.id;
+                const minQty = combo.qty_free;
+                const totalQtyAvailable = combo.combo_item_ids.reduce((sum, item) => {
+                    const productId = item.product_id.id;
+                    if (availableQty[productId]) {
+                        return (
+                            sum +
+                            Object.values(availableQty[productId]).reduce(
+                                (sum, lineQty) => sum + lineQty,
+                                0
+                            )
+                        );
+                    }
+                    return sum;
+                }, 0);
+                quantityTaken[comboId] = {};
+
+                if (totalQtyAvailable < minQty && !combo.is_upsell) {
+                    // Not enough to satisfy qty_free for this combo group
+                    comboQty = 0;
+                    break;
+                }
+
+                if (totalQtyAvailable >= minQty) {
+                    comboQty = Math.min(totalQtyAvailable / minQty, comboQty);
+                }
+            }
+
+            if (comboQty === 0) {
+                continue;
+            }
+
+            if (comboQty == Infinity) {
+                // The combo product is composed of only upsell combo choices
+                // We do not propose it as applicable or upsell
+                continue;
+            }
+
+            const combinations = [];
+            for (let i = 0; i < comboQty; i++) {
+                const quantityTaken = {};
+                for (const combo of combos) {
+                    const comboId = combo.id;
+                    const totalQtyAvailable = combo.combo_item_ids.reduce((sum, item) => {
+                        const productId = item.product_id.id;
+                        if (availableQty[productId]) {
+                            return (
+                                sum +
+                                Object.values(availableQty[productId]).reduce(
+                                    (sum, lineQty) => sum + lineQty,
+                                    0
+                                )
+                            );
+                        }
+                        return sum;
+                    }, 0);
```

Comment:
This is done two times, above and here. You can precompute this data above both loops.